storage/src/journal/fixed.rs

```diff
@@ -0,0 +1,656 @@
+    ///
+    /// All backing blobs are opened but not read during initialization. The `replay` method can be
+    /// used to iterate over all items in the `Journal`.
+    pub async fn init(runtime: E, cfg: Config) -> Result<Self, Error> {
+        // Iterate over blobs in partition
+        let mut blobs: Vec<(B, u64)> = Vec::new();
+        let stored_blobs = match runtime.scan(&cfg.partition).await {
+            Ok(blobs) => blobs,
+            Err(RError::PartitionMissing(_)) => Vec::new(),
+            Err(err) => return Err(Error::Runtime(err)),
+        };
+        for name in stored_blobs {
+            let blob = runtime
+                .open(&cfg.partition, &name)
+                .await
+                .map_err(Error::Runtime)?;
+            let index = match name.try_into() {
+                Ok(index) => u64::from_be_bytes(index),
+                Err(nm) => return Err(Error::InvalidBlobName(hex(&nm))),
+            };
+            debug!(blob = index, "loaded blob");
+            blobs.push((blob, index));
+        }
+        blobs.sort_by(|(_, a), (_, b)| a.cmp(b));
+        let mut oldest_blob = 0;
+        if !blobs.is_empty() {
+            // Check that there are no gaps in the blob numbering, which would indicate missing data.
+            let mut it = blobs.iter();
+            let mut previous_blob = it.next().unwrap().1;
+            oldest_blob = previous_blob;
+            for (_, section) in it {
+                if *section != previous_blob + 1 {
+                    return Err(Error::MissingBlob(previous_blob + 1));
+                }
+                previous_blob = *section;
+            }
+        } else {
+            debug!("no blobs found, bootstrapping with 1 empty blob");
+            let blob = runtime.open(&cfg.partition, &0u64.to_be_bytes()).await?;
+            blobs.push((blob, 0));
+        }
+
+        // Initialize metrics
+        let tracked = Gauge::default();
+        let synced = Counter::default();
+        let pruned = Counter::default();
+        {
+            let mut registry = cfg.registry.lock().unwrap();
+            registry.register("tracked", "Number of blobs", tracked.clone());
+            registry.register("synced", "Number of syncs", synced.clone());
+            registry.register("pruned", "Number of blobs pruned", pruned.clone());
+        }
+        tracked.set(blobs.len() as i64);
+
+        let blobs: Vec<B> = blobs.into_iter().map(|(blob, _)| blob).collect();
+        // truncate the last blob if it's not the expected length, which might happen from unclean shutdown.
+        let last_blob = blobs.last().unwrap();
+        let blob_len: u64 = last_blob.len().await?;
+        if blob_len % (size_of::<u32>() + N) as u64 != 0 {
+            warn!(
+                "last blob len ({}) is not a multiple of item size, truncating",
+                blob_len
+            );
+            last_blob
+                .truncate(blob_len - blob_len % (size_of::<u32>() + N) as u64)
```

Comment:
nit: recommend breaking `blob_len % (size_of::<u32>() + N)` into a var you can reuse.

Reply:
done